File: ColoringBookApp/Features/Canvas/RegionMask.swift
import CoreGraphics
import UIKit

/// Decoded region label image. Each closed line-art region has a unique RGB color.
/// Tapping a point looks up its region color, then we recolor every pixel sharing
/// that label in the paint layer — O(N) once, then cached.
final class RegionMask {
    let width: Int
    let height: Int
    private let pixels: [UInt32]   // BGRA packed
    private var regionPixelIndex: [UInt32: [Int]] = [:]   // labelColor -> linear pixel indices

    init?(image: UIImage) {
        guard let cg = image.cgImage else { return nil }
        let w = cg.width
        let h = cg.height
        var buffer = [UInt32](repeating: 0, count: w * h)
        let cs = CGColorSpaceCreateDeviceRGB()
        let info = CGImageAlphaInfo.premultipliedFirst.rawValue
            | CGBitmapInfo.byteOrder32Little.rawValue
        guard let ctx = CGContext(
            data: &buffer,
            width: w, height: h,
            bitsPerComponent: 8, bytesPerRow: w * 4,
            space: cs, bitmapInfo: info
        ) else { return nil }
        ctx.draw(cg, in: CGRect(x: 0, y: 0, width: w, height: h))
        self.width = w
        self.height = h
        self.pixels = buffer
    }

    /// Region label (RGB) at the given normalized coordinate (0...1).
    /// Returns nil if outside or on a line-art (black) pixel.
    func regionLabel(atNormalized point: CGPoint) -> UInt32? {
        guard point.x >= 0, point.x <= 1, point.y >= 0, point.y <= 1 else { return nil }
        let x = min(width - 1, max(0, Int(point.x * CGFloat(width))))
        let y = min(height - 1, max(0, Int(point.y * CGFloat(height))))
        let p = pixels[y * width + x]
        // ignore line-art (near-black) and transparent
        let r = (p >> 16) & 0xFF
        let g = (p >> 8) & 0xFF
        let b = p & 0xFF
        let a = (p >> 24) & 0xFF
        if a < 8 { return nil }
        if r < 24 && g < 24 && b < 24 { return nil }
        return (r << 16) | (g << 8) | b
    }

    /// Indices of every pixel with this label, indexed lazily.
    func pixelIndices(forLabel label: UInt32) -> [Int] {
        if let cached = regionPixelIndex[label] { return cached }
        var out: [Int] = []
        out.reserveCapacity(width * height / 32)
        for i in 0..<pixels.count {
            let p = pixels[i]
            let r = (p >> 16) & 0xFF
            let g = (p >> 8)  & 0xFF
            let b = p & 0xFF
            let labeled = (r << 16) | (g << 8) | b
            if labeled == label { out.append(i) }
        }
        regionPixelIndex[label] = out
        return out
    }
}

File: web/js/region-mask.js
// Tap-to-fill core. Mirrors iOS RegionMask + PaintLayer.
// Each page ships with a region label image where every closed line-art
// region has a unique RGB. A tap maps to a region label, then we recolor
// every pixel sharing that label in the paint layer in O(N), cached so a
// repeat tap on the same region is O(1).

export class RegionMask {
  /**
   * @param {ImageData} imageData The region label image as raw pixels.
   */
  constructor(imageData) {
    this.width = imageData.width;
    this.height = imageData.height;
    this.data = imageData.data; // Uint8ClampedArray, RGBA
    this._indexCache = new Map(); // labelKey -> Uint32Array of pixel indices
  }

  /**
   * Look up the label color at a normalized point (0..1).
   * Returns null on transparent or near-black (line-art) pixels.
   */
  labelAtNormalized(x, y) {
    if (x < 0 || x > 1 || y < 0 || y > 1) return null;
    const px = Math.min(this.width - 1, Math.max(0, Math.floor(x * this.width)));
    const py = Math.min(this.height - 1, Math.max(0, Math.floor(y * this.height)));
    const i = (py * this.width + px) * 4;
    const r = this.data[i], g = this.data[i + 1], b = this.data[i + 2], a = this.data[i + 3];
    if (a < 8) return null;
    if (r < 24 && g < 24 && b < 24) return null;
    return (r << 16) | (g << 8) | b;
  }

  /**
   * Indices into the paint layer's Uint32 buffer for every pixel matching a
   * given label, computed lazily and cached.
   */
  pixelIndices(label) {
    const cached = this._indexCache.get(label);
    if (cached) return cached;
    const total = this.width * this.height;
    const out = new Uint32Array(total);
    let n = 0;
    const targetR = (label >> 16) & 0xff;
    const targetG = (label >> 8) & 0xff;
    const targetB = label & 0xff;
    for (let i = 0; i < total; i++) {
      const j = i * 4;
      if (this.data[j] === targetR && this.data[j + 1] === targetG && this.data[j + 2] === targetB) {
        out[n++] = i;
      }
    }
    const indices = out.slice(0, n);
    this._indexCache.set(label, indices);
    return indices;
  }
}

/** PaintLayer: a Uint32 buffer behind the line-art layer. */
export class PaintLayer {
  constructor(width, height) {
    this.width = width;
    this.height = height;
    this.imageData = new ImageData(width, height);
    this.u32 = new Uint32Array(this.imageData.data.buffer);
  }

  /** Fill every pixel at `indices` with the given [r,g,b,a]. Returns prior values. */
  fill(indices, rgba) {
    const packed = pack(rgba);
    const prev = new Uint32Array(indices.length);
    for (let k = 0; k < indices.length; k++) {
      const i = indices[k];
      prev[k] = this.u32[i];
      this.u32[i] = packed;
    }
    return prev;
  }

  restore(indices, prev) {
    for (let k = 0; k < indices.length; k++) {
      this.u32[indices[k]] = prev[k];
    }
  }

  clear() {
    this.u32.fill(0);
  }

  replace(u32) {
    if (u32.length === this.u32.length) this.u32.set(u32);
  }

  draw(ctx) {
    ctx.putImageData(this.imageData, 0, 0);
  }
}

/** Convert [r,g,b,a] (0..255) to a little-endian RGBA Uint32 (the layout used by ImageData on every modern engine). */
function pack(rgba) {
  return ((rgba[3] & 0xff) << 24) | ((rgba[2] & 0xff) << 16) | ((rgba[1] & 0xff) << 8) | (rgba[0] & 0xff);
}
